node/lib/stitch.js

```diff
@@ -0,0 +1,243 @@
+    help: `submodules whose paths are matched by this regex are not stitched, \
+but are instead kept as submodules.`,
+    required: false,
+    defaultValue: null,
+});
+
+parser.addArgument(["--skip-empty"], {
+    required: false,
+    action: "storeConst",
+    constant: true,
+    defaultValue: false,
+    help: "Skip a commit if its  tree would be the same as its first parent.",
+});
+
+parser.addArgument(["--root"], {
+    type: "string",
+    help: "When provided, run a *join* operation that creates a new history \
+joining those of the submodules under the specified 'root' path.  \
+In this history those paths will be relative to 'root', i.e., they will not \
+be prefixed by it. This option implies '--skip-empty'.",
+});
+
+const stitch = co.wrap(function *(repoPath,
+                                  commitish,
+                                  url,
+                                  keepAsSubmodule,
+                                  joinRoot,
+                                  numParallel,
+                                  fetch,
+                                  targetBranchName,
+                                  skipEmpty) {
+    const repo = yield NodeGit.Repository.open(repoPath);
+    const annotated = yield GitUtil.resolveCommitish(repo, commitish);
+    if (null === annotated) {
+        throw new Error(`Could not resolve ${commitish}.`);
+    }
+    const commit = yield repo.getCommit(annotated.id());
+
+    console.log("listing unconverted ancestors of", commit.id().tostrS());
+
+    const commitsToStitch = yield StitchUtil.listCommitsToStitch(repo, commit);
+
+    const adjustPath = StitchUtil.makeAdjustPathFunction(joinRoot);
+
+    console.log(commitsToStitch.length, "to stitch");
+
+    if (fetch) {
+        console.log("listing fetches");
+        const fetches = yield StitchUtil.listFetches(repo,
+                                                     commitsToStitch,
+                                                     keepAsSubmodule,
+                                                     adjustPath,
+                                                     numParallel);
+        console.log("Found", Object.keys(fetches).length, "subs to fetch.");
+        const subNames = Object.keys(fetches);
+        const doFetch = co.wrap(function *(name, i) {
+            const subFetches = fetches[name];
+            const fetchTimeMessage = `\
+(${i + 1}/${subNames.length}) -- fetched ${subFetches.length} SHAs for \
+${name}`;
+            console.time(fetchTimeMessage);
+            yield StitchUtil.fetchSubCommits(repo, url, subFetches);
+            console.timeEnd(fetchTimeMessage);
+        });
+        yield DoWorkQueue.doInParallel(subNames, doFetch, numParallel);
+    }
+
+    console.log("Now stitching");
+    let lastCommit = null;
+
+    for (let i = 0; i < commitsToStitch.length; ++i) {
+        const next = commitsToStitch[i];
+
+        // If we had the same commit in the graph more than once, don't convert
```

Comment:
This does not actually seem to check that right here?

Reply:
Yes, looks like that comment is dangling.